### DISCUSSION TITLE
Deploy Next.JS app with O2Switch's CPanel ISSUE

### DISCUSSION THREAD

**Discussion by Benja-san** [vercel/next.js#32001] [Help]
Hello Team,
I am currently deploying a test app on my client's server based on 02Switch. Unfortunately it seems a bit complicated to deploy node apps with this host.

I managed successfully to get the ssh connection, I cloned the repo onto the server but when I try to build the app, I have this error coming up : 

`error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error

> Build error occurred
RangeError: WebAssembly.instantiate(): Out of memory: wasm memory
    at internal/deps/cjs-module-lexer/dist/lexer.js:1:33593
    at async initCJSParse (internal/modules/esm/translators.js:72:5)
    at async Loader.commonjsStrategy (internal/modules/esm/translators.js:185:18)
    at async link (internal/modules/esm/module_job.js:67:21)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! _app-name_@ build: `next build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the  _app-name_@ build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/vnzg2266/.npm/_logs/2021-12-01T10_53_56_837Z-debug.log`

Has someone got the same trouble, trying to build a node app onto a server? And if yes, could you help me :D ??

Thanks a lot for taking time reading it!

**Comment by slobodanmilinkovic** [comment]
I have the exact same problem. Someone, please.

**Reply by aboda1986** [reply]
same issue for me

**Comment by ActivatedDigital** [comment]
+1

**Comment by asheroto** [comment]
Same here...

**Comment by hellotreedigital** [comment]
SAME :(

**Comment by amaurylebouteux** [comment]
Same ...

**Comment by amaurylebouteux** [comment]
Same ...
this worked for me: https://dev.to/techfortified/how-to-host-nextjs-website-on-cpanel-2dij

**Comment by onlinejo** [comment]
hello some news ?

**Comment by loicrey** [comment]
Are you using the terminal inside CPanel from your browser ?

**Reply by onlinejo** [reply]
yes

**Reply by loicrey** [reply]
The browser Terminal have some limitations. Consider trying with a classic Terminal and SSH connection to your o2switch server
see : https://faq.o2switch.fr/hebergement-mutualise/tutoriels-cpanel/terminal#erreur-de-memoire-nodenpm

**Reply by asheroto** [reply]
Not for me, running on local, haven't tested in awhile

**Comment by mcasters** [comment]
"Out of memory: wasm memory" : There is not enough resources to build the project.
Build in local, and copy the .next folder in your production root app.

**Comment by onlinejo** [comment]
i  think o2switch is not good hosting for  nextJS , they use old technology

**Reply by JayMds** [reply]
This is not Next.Js based issue and i'm not even Next.js user, by the way i solved my problem hosting node app on O2switch, the terminal from cpanel is memory limited, you have to login there with SSH (white list your ip 1st), once you're logged use the 'source' cmd specified in node.js app panel of your project, and you could build your app from there

https://faq.o2switch.fr/hebergement-mutualise/tutoriels-cpanel/app-nodejs
from their doc if you read "deploying react app"  you will usefull informations but you will have to adapt for your case, also support could give you more informations

**Comment by JayMds** [comment]
This is not Next.Js based issue and i'm not even Next.js user, by the way i solved my problem hosting node app on O2switch, the terminal from cpanel is memory limited, you have to login there with SSH (white list your ip 1st), once you're logged use the 'source' cmd specified in node.js app panel of your project, and you could build your app from there

https://faq.o2switch.fr/hebergement-mutualise/tutoriels-cpanel/app-nodejs
from their doc if you read "deploying react app" you will usefull informations but you will have to adapt for your case, also support could give you more informations

**Comment by sohelzerdoumi** [comment]
For o2switch deployment,

# Packages

install package from ssh to bypass memory limits and webassembly:

```bash
source /home/myaccount/nodevenv/app/myapplication/release/20/bin/activate
cd /home/myaccount/app/myapplication/release
npm install
```

# Prisma

for prisma, change your schema binaryTargets with correct ssl version

```schema.prisma
generator client {
provider = "prisma-client-js"
binaryTargets = ["native", "debian-openssl-1.0.x"]
previewFeatures = []
}
```

generate prisma assets:
```bash
npm exec prisma generate
```

# Logs

in order to have production logs, update .htaccess with:
```.htaccess
PassengerAppLogFile "/home/myaccount/app/myapplication/release/production.log"
```

All of this worked for me

**Comment by cedricrgt** [comment]
I'm currently struggling with deploying a Next.js application on o2switch. While all of my static pages are deployed and working perfectly, my dynamic routes are failing consistently.

I can confirm that the backend is properly sending the requested data, but any dynamic ID page immediately returns a 500 Internal Server Error on the frontend. Here is an example of a failing route: http://sc1yefa3951.universe.wf/details/0ba9084e-d38d-4986-856e-ddb85acd8a26

Given the known complexities of running Next.js via Phusion Passenger on shared hosting, does anyone have an updated solution or .htaccess configuration to prevent dynamic routes from crashing with a 500 error?

**Comment by alexandre-leng** [comment]
same issue

**Comment by RyanH3-DEV** [comment]
I'm considering hosting a Next.js project (frontend + API) with PostgreSQL on o2switch. I know it had some limitations with Node.js in the past. Has the experience improved recently? Is it stable enough for production, and would you recommend it over a VPS or other hosting providers?

**Reply by cedricrgt** [reply]
Hello, I ve managed to do it. I have struggle with the IP white list as it is limited at 5 up you will have to use a script to clear the list automatically at every push. Otherwise it is pretty stable, depends on traffic I guess. I haven't done a loading test and it was a school project.
http://sc1yefa3951.universe.wf/
